### ISSUE TITLE
Buffers are reused due to inlining of `runBuffer`

### ISSUE THREAD

**Issue by KyleDavidE** [Bodigrim/linear-builder#19]
`runBuffer (\b -> b |>. '_' |>. 'a' |> runBuffer (\b' -> b' |>. '_' |>. 'b'))` Outputs "_b_b" when the expected output is "_a_b"

I think this is because runBuffer can remove common subexpressions in bad places. So it finds the common subexpression of `Buffer mempty |>. '_'` and re-uses it.

Also the worker functions generated for the functions provided by this package aren't linear because they operate over unpacked text, I don't know if that has a role to play here.

**Comment by Bodigrim** [comment]
@KyleDavidE thanks, this is very interesting. 

Is it possible to trigger a similar bug with [`Data.Array.Mutable.Linear`](https://hackage.haskell.org/package/linear-base-0.4.0/docs/Data-Array-Mutable-Linear.html)? In other words: is it me being stupid or is GHC performing CSE in the middle of a linear pipeline at fault?

CC @aspiwack

**Comment by KyleDavidE** [comment]
I think GHC is in the right here.

Wrt to [Data.Array.Mutable.Linear](https://hackage.haskell.org/package/linear-base-0.4.0/docs/Data-Array-Mutable-Linear.html) I don't think the same thing would work there because `Data.Array.Mutable.Unlifted.Linear.Unlifted.alloc` is noinline.

I think marking runBuffer as noInline would probably also work, but I'm not sure. A weirder idea that might also work would be putting a state token inside of `Buffer` but I'm not sure if that is safe.


Messing with it a bit more:

```hs
{-# NOINLINE foo #-}
foo :: (Buffer ⊸ Buffer) → (Buffer ⊸ Buffer) → (Buffer ⊸ Buffer) → (Buffer ⊸ Text → Buffer) → Text
foo f g g' app = runBuffer \b → (\fb -> (g fb) `app` runBuffer (\b -> g' (f b))) (f b)


main = print (foo (|>. '_') (|>. 'a') (|>. 'b') (|>))
```

also repros the issue.

What I think is going on here is that after runBuffer inlines it isn't a linear pipeline anymore. Linear functions are not functions that must be used linearly, they are functions that 

The simplifier gives for foo:
```
-- RHS size: {terms: 67, types: 110, coercions: 8, joins: 0/0}
MyLib.$wfoo
  :: (Buffer %1 -> Buffer)
     -> (Buffer %1 -> Buffer)
     -> (Buffer %1 -> Buffer)
     -> (Buffer %1 -> Text -> Buffer)
     -> (# ghc-prim:GHC.Prim.ByteArray#, ghc-prim:GHC.Prim.Int#,
           ghc-prim:GHC.Prim.Int# #)
MyLib.$wfoo
  = \ (f_s2gp :: Buffer %1 -> Buffer)
      (g_s2gq :: Buffer %1 -> Buffer)
      (g'_s2gr :: Buffer %1 -> Buffer)
      (app_s2gs :: Buffer %1 -> Text -> Buffer) ->
      case Data.Text.Array.empty of
      { Data.Array.Byte.ByteArray unbx_i1Sm ->
      case f_s2gp (Data.Text.Builder.Linear.Core.Buffer unbx_i1Sm 0# 0#)
      of fb_a1xm
      { Data.Text.Builder.Linear.Core.Buffer ipv_s1Tr ipv1_s1Ts
                                             ipv2_s1Tt ->
      case app_s2gs
             (g_s2gq fb_a1xm)
             (case g'_s2gr fb_a1xm of
              { Data.Text.Builder.Linear.Core.Buffer ipv3_i1T8 ipv4_i1T9
                                                     ipv5_i1Ta ->
              ghc-prim:GHC.Magic.runRW#
                (\ (s_i1SG
                      :: ghc-prim:GHC.Prim.State# ghc-prim:GHC.Prim.RealWorld) ->
                   case Unsafe.Coerce.unsafeEqualityProof of
                   { Unsafe.Coerce.UnsafeRefl v2_i1T7 ->
                   case ghc-prim:GHC.Prim.shrinkMutableByteArray#
                          (ipv3_i1T8
                           `cast` <Co:2> :: ghc-prim:GHC.Prim.ByteArray#
                                            ~R# ghc-prim:GHC.Prim.MutableByteArray#
                                                  ghc-prim:GHC.Prim.RealWorld)
                          (ghc-prim:GHC.Prim.+# ipv4_i1T9 ipv5_i1Ta)
                          s_i1SG
                   of s2#_i1Td
                   { __DEFAULT ->
                   case ghc-prim:GHC.Prim.unsafeFreezeByteArray#
                          (ipv3_i1T8
                           `cast` <Co:2> :: ghc-prim:GHC.Prim.ByteArray#
                                            ~R# ghc-prim:GHC.Prim.MutableByteArray#
                                                  ghc-prim:GHC.Prim.RealWorld)
                          s2#_i1Td
                   of
                   { (# ipv6_i1Tf, ipv7_i1Tg #) ->
                   Data.Text.Internal.Text ipv7_i1Tg ipv4_i1T9 ipv5_i1Ta
                   }
                   }
                   })
              })
      of
      { Data.Text.Builder.Linear.Core.Buffer ipv3_i1T8 ipv4_i1T9
                                             ipv5_i1Ta ->
      ghc-prim:GHC.Magic.runRW#
        (\ (s_i1SG
              :: ghc-prim:GHC.Prim.State# ghc-prim:GHC.Prim.RealWorld) ->
           case Unsafe.Coerce.unsafeEqualityProof of
           { Unsafe.Coerce.UnsafeRefl v2_i1T7 ->
           case ghc-prim:GHC.Prim.shrinkMutableByteArray#
                  (ipv3_i1T8
                   `cast` <Co:2> :: ghc-prim:GHC.Prim.ByteArray#
                                    ~R# ghc-prim:GHC.Prim.MutableByteArray#
                                          ghc-prim:GHC.Prim.RealWorld)
                  (ghc-prim:GHC.Prim.+# ipv4_i1T9 ipv5_i1Ta)
                  s_i1SG
           of s2#_i1Td
           { __DEFAULT ->
           case ghc-prim:GHC.Prim.unsafeFreezeByteArray#
                  (ipv3_i1T8
                   `cast` <Co:2> :: ghc-prim:GHC.Prim.ByteArray#
                                    ~R# ghc-prim:GHC.Prim.MutableByteArray#
                                          ghc-prim:GHC.Prim.RealWorld)
                  s2#_i1Td
           of
           { (# ipv6_i1Tf, ipv7_i1Tg #) ->
           (# ipv7_i1Tg, ipv4_i1T9, ipv5_i1Ta #)
           }
           }
           })
      }
      }
      }
```

**Comment by KyleDavidE** [comment]
Another repo without the nesting:

```hs
bar :: (Buffer ⊸ Buffer) → (Buffer ⊸ Buffer)  → (Text, Text)
bar g f = 
    let !a = runBuffer \b -> g (f b |>. 'a')
        !b = runBuffer \b -> g (f b |>. 'b')
    in (a,b)

main = print (bar (|>… 5) (|>. '_'))
```

outputs `("_a     ","_a     ")`

to be clear bar doesn't need to be abstracted out like this, it just makes the simplifier output clearer.

**Comment by Bodigrim** [comment]
Thanks again, @KyleDavidE. I pushed 798670d, it seems to make things better. Could you please give a try? If it looks good, I'll make a release soon.

**Comment by aspiwack** [comment]
It is all really subtle indeed. It took us several iterations until we were confident that the API was safe. At its core the problem stems from `unsafePerformIO`, and we need well-placed `NOINLINE` pragmas. But there are additional complications, I don't quite remember the saga, but we've had some laziness problems too.

**Comment by Bodigrim** [comment]
@facundominguez @utdemir might you be interested to take a look at this? Is 798670ddac2ac70cf43aadc36c2980f0b481d099 sufficient to fix sharing of mutable linear buffers? Or am I missing something else?

**Comment by Bodigrim** [comment]
Fixed and released as [`text-builder-linear-0.1.2`](https://hackage.haskell.org/package/text-builder-linear-0.1.2), thanks all.
